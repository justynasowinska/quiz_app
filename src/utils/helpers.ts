import { AllHtmlEntities as Entities } from 'html-entities';
import { DifficultlyType } from '../redux/types';
import { THEME_COLORS } from '../config/colors';

const entities = new Entities();

/**
 * Replaces entities to characters. Unknown entities are left as is.
 * @param {string} text     Text to decode.
 * @returns Text with entities converted to text.
 */
export const decodeWithEntities = (text: string): string => entities.decode(text);

/**
 * Gets color with regard to given difficulty level.
 * @param {DifficultlyType} difficulty  Difficulty level.
 * @returns Color for given difficulty.
 */
export const getDifficultyLevelColor = (difficulty: DifficultlyType): string => {
    switch (difficulty) {
        case 'easy':
            return THEME_COLORS.easyQuestionLevel;
        case 'medium':
            return THEME_COLORS.mediumQuestionLevel;
        case 'hard':
            return THEME_COLORS.hardQuestionLevel;
    }
};