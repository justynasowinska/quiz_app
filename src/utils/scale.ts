import { SCREEN_WIDTH, SCREEN_HEIGHT } from './metrics';

const guidelineBaseWidth = 375;
const guidelineBaseHeight = 667;

interface ScaleType {
    width?: number;
    size: number;
}

const scale = ({ size, width = SCREEN_WIDTH }: ScaleType) => (width / guidelineBaseWidth) * size;

interface VerticalScaleType {
    height?: number;
    size: number;
}

const verticalScale = ({ size, height = SCREEN_HEIGHT }: VerticalScaleType) => (height / guidelineBaseHeight) * size;

interface ModerateScaleType {
    size: number;
    width?: number;
    factor?: number;
    }

const moderateScale = ({
    size,
    width = SCREEN_WIDTH,
    factor = 0.5,
}: ModerateScaleType) => size + (scale({ width, size }) - size) * factor;

interface ModerateVerticalScaleType {
    size: number;
    height?: number;
    factor?: number;
}

const moderateVerticalScale = ({
    size,
    height = SCREEN_HEIGHT,
    factor = 0.5,
}: ModerateVerticalScaleType) =>
    size + (verticalScale({ height, size }) - size) * factor;

export {
    scale,
    verticalScale,
    moderateScale,
    moderateVerticalScale,
};