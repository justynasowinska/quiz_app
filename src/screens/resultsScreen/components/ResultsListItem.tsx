import * as React from 'react';
import { View, Text, StyleSheet } from 'react-native';
import { Icon } from 'react-native-elements';

import { THEME_COLORS } from '../../../config/colors';
import { decodeWithEntities } from '../../../utils/helpers';
import { setTestID } from '../../../utils/TestingUtils';

export interface ResultType {
    questionId: number;
    question: string;
    correct: boolean;
    correctAnswer: string;
}

interface PropsType {
    resultsItem: ResultType;
}

export const ResultsListItem = (props: PropsType) => {
    const renderIcon = (correct: boolean) => {
        if (correct) {
            return (
                <Icon
                    name="check"
                    color={THEME_COLORS.green}
                    iconStyle={styles.iconStyle}
                />
            );
        }

        return (
            <Icon
                name="close"
                color={THEME_COLORS.red}
                iconStyle={styles.iconStyle}
            />
        );
    }

    const { resultsItem: { question, correct, correctAnswer } } = props;
    return (
        <View style={styles.container}>
            {renderIcon(correct)}
            <View style={styles.textContainer}>
                <Text style={styles.questionText}>{decodeWithEntities(question)}</Text>
                { !correct && <Text testID={setTestID('results_correct_answer')} style={styles.correctAnswer}>{`Correct answer: ${correctAnswer}`}</Text> }
            </View>
        </View>
    );
};

const styles = StyleSheet.create({
    container: {
        flexDirection: 'row',
        marginTop: 10,
        marginBottom: 10,
        alignItems: 'center',
    },
    textContainer: {
        flexDirection: 'column',
        flexShrink: 1,
    },
    questionText: {
        fontSize: 20,
        color: THEME_COLORS.appTextColor
    },
    correctAnswer: {
        fontSize: 15,
        color: THEME_COLORS.green
    },
    iconStyle: {
        marginRight: 10
    }
});