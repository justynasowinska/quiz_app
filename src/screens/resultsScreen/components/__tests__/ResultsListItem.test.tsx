import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';

import { ResultsListItem } from '../ResultsListItem';
import { findByTestID } from '../../../../utils/TestingUtils';

describe('Results Screen ResultsListItem Component', () => {
    let component: ShallowWrapper<any, Readonly<{}>, React.Component<{}, {}, any>>;

    const results = {
        questionId: 0,
        question: 'Unturned originally started as a Roblox game.',
        correct: true,
        correctAnswer: 'True'
    };

    beforeEach(() => {
        component = shallow(
            <ResultsListItem
                resultsItem={results}
            />
        );
    });

    it('matches to the snapshot', () => {
        expect(component).toMatchSnapshot();

        component.setProps({ resultsItem: { ...results, correct: false } });
        expect(component).toMatchSnapshot();
    });

    it('renders correct answer only when user\'s answer was wrong', () => {
        component.setProps({ resultsItem: { ...results, correct: false } });

        expect(findByTestID(component, 'results_correct_answer')).toHaveLength(1);

        component.setProps({ resultsItem: { ...results, correct: true } });

        expect(findByTestID(component, 'results_correct_answer')).toHaveLength(0);
    });
});
