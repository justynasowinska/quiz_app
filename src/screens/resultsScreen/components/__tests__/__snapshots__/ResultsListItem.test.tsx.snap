// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Results Screen ResultsListItem Component matches to the snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "marginBottom": 10,
      "marginTop": 10,
    }
  }
>
  <Themed.Icon
    color="#2FA83E"
    iconStyle={
      Object {
        "marginRight": 10,
      }
    }
    name="check"
  />
  <View
    style={
      Object {
        "flexDirection": "column",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#474D54",
          "fontSize": 20,
        }
      }
    >
      Unturned originally started as a Roblox game.
    </Text>
  </View>
</View>
`;

exports[`Results Screen ResultsListItem Component matches to the snapshot 2`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "marginBottom": 10,
      "marginTop": 10,
    }
  }
>
  <Themed.Icon
    color="#CD0720"
    iconStyle={
      Object {
        "marginRight": 10,
      }
    }
    name="close"
  />
  <View
    style={
      Object {
        "flexDirection": "column",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#474D54",
          "fontSize": 20,
        }
      }
    >
      Unturned originally started as a Roblox game.
    </Text>
    <Text
      style={
        Object {
          "color": "#2FA83E",
          "fontSize": 15,
        }
      }
      testID="results_correct_answer"
    >
      Correct answer: True
    </Text>
  </View>
</View>
`;
